core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -538,13 +541,22 @@ class ReplicaManager(val config: KafkaConfig,
           val initialLogEndOffset = localReplica.logEndOffset
           val logReadInfo = localReplica.log match {
             case Some(log) =>
-              val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
+              val adjustedFetchSize =
+                if (Topic.isInternal(topic) && !readOnlyCommitted)
+                  Math.max(fetchSize, log.config.maxMessageSize)
+                else if (limitBytes > 0)
+                  fetchSize
+                else
+                  0
               log.read(offset, adjustedFetchSize, maxOffsetOpt)
             case None =>
               error("Leader for partition [%s,%d] does not have a local log".format(topic, partition))
               FetchDataInfo(LogOffsetMetadata.UnknownOffsetMetadata, MessageSet.Empty)
           }
 
+          if (limitBytes != Int.MaxValue)
```

Comment:
Yes, I understood that compatibility was the intent, but I was trying to figure out how it would behave differently without the check. You're thinking of cases where the response size would be larger than 2^31?